Python_distribution example

lgvital opened the pull request:
Beginnings of #4740 

### Problem

There is no way to consume a python_library with c_extensions.

### Solution

Create a new target, `python_distribution` that takes in a setup.py. Create an example of how this might be done from a UX standpoint. We've run a similar example by our customer to validate that this meets their needs.

### Result

None yet, mainly documentation and will be used as a fixture for an integration test in the future. Running pex on the new superhello/ directory runs successfully. We will basically be doing the code equivalent in a task.

kwlzn commented on examples/src/python/example/python_distribution/hello/main/BUILD:
```diff
@@ -0,0 +1,11 @@
+# Copyright 2014 Pants project contributors (see CONTRIBUTORS.md).
```
these should bump to 2017.

stuhood commented:
@kwlzn : Do you think it would be feasible to ask to have python modules explicitly depend on `c(pp)_library` targets, rather than embedding the code? Would require pulling that code into core. But maybe the python toolchain is handling it to a degree where having another compiler involved would be redundant...?

kwlzn commented:
> @kwlzn : Do you think it would be feasible to ask to have python modules explicitly depend on c(pp)_library targets, rather than embedding the code? Would require pulling that code into core. But maybe the python toolchain is handling it to a degree where having another compiler involved would be redundant...?

feasible, yes - and the ideal long term strategy IMHO particular around python + rust.. but I don't think its super practical for the short term (i.e. needed this quarter). `c(pp)_library` afaik isn't super mature (or used) and would require a bit of custom integration work. `python_distribution` nets us a much simpler initial approach /and/ buys us python dist interop, which is a win/win in my book even if we do eventually implement cross-target interop between c/cpp/rust/python et al.

lgvital commented:
Thanks, will remove the excess __init__.py's. Will definitely be landing this alongside the actual task/target changes. Pairing up with @CMLivingston on this work--what's the preferred method for dependent changes for pants? Branch off branch -> pull request for each -> merge all at the same time? One large pull request?

> feasible, yes - and the ideal long term strategy IMHO particular around python + rust.. but I don't think its super practical for the short term (i.e. needed this quarter). c(pp)_library afaik isn't super mature (or used) and would require a bit of custom integration work. python_distribution nets us a much simpler initial approach /and/ buys us python dist interop, which is a win/win in my book even if we do eventually implement cross-target interop between c/cpp/rust/python et al.

Agreed. Short-term, python_distribution is a practical, not too difficult way to go to unblock internal customers. Long-term, better support for c(pp)_library targets.

kwlzn commented:
> what's the preferred method for dependent changes for pants? Branch off branch -> pull request for each -> merge all at the same time? One large pull request?

we try to stick to around <=600 line PRs, so for a new change like this that might be 1-3 initial 100-600 line PRs + some iterative follow-ons/bugfixes. since you can't merge them yourself, it's probably most ideal to focus on medium to large PRs vs feathering a bunch of tiny ones.

lgvital commented:
Closing this in favor of #5141 (WIP), which tweaks the example, addresses comments here, implements target and a modifies `python_binary_create` task to handle `python_distribution` targets.
